vcd/resource_vcd_vapp_firewall_rules.go

```diff
@@ -0,0 +1,394 @@
+	vcdClient := meta.(*VCDClient)
+
+	_, vdc, err := vcdClient.GetOrgAndVdcFromResource(d)
+	if err != nil {
+		return fmt.Errorf(errorRetrievingOrgAndVdc, err)
+	}
+
+	vappId := d.Get("vapp_id").(string)
+	vapp, err := vdc.GetVAppById(vappId, false)
+	if err != nil {
+		return fmt.Errorf("error finding vApp. %s", err)
+	}
+
+	vcdClient.lockParentVappWithName(d, vapp.VApp.Name)
+	defer vcdClient.unLockParentVappWithName(d, vapp.VApp.Name)
+
+	_, err = vapp.UpdateNetworkFirewallRules(d.Get("network_id").(string), []*types.FirewallRule{},
+		d.Get("default_action").(string), d.Get("log_default_action").(bool))
+	if err != nil {
+		log.Printf("[INFO] Error setting firewall rules: %s", err)
+		return fmt.Errorf("error setting firewall rules: %#v", err)
+	}
+
+	return nil
+}
+
+func resourceVappFirewallRulesRead(d *schema.ResourceData, meta interface{}) error {
+	vcdClient := meta.(*VCDClient)
+
+	_, vdc, err := vcdClient.GetOrgAndVdcFromResource(d)
+	if err != nil {
+		return fmt.Errorf(errorRetrievingOrgAndVdc, err)
+	}
+
+	vappId := d.Get("vapp_id").(string)
+	vapp, err := vdc.GetVAppById(vappId, false)
+	if err != nil {
+		return fmt.Errorf("error finding vApp. %s", err)
+	}
+
+	vappNetwork, err := vapp.GetVappNetworkById(d.Get("network_id").(string), false)
+	if err != nil {
+		return fmt.Errorf("error finding vApp network. %s", err)
+	}
+
+	var rules []map[string]interface{}
+	for _, rule := range vappNetwork.Configuration.Features.FirewallService.FirewallRule {
+		singleRule := make(map[string]interface{})
+		singleRule["description"] = rule.Description
+		singleRule["enabled"] = rule.IsEnabled
+		singleRule["policy"] = rule.Policy
+		singleRule["protocol"] = getProtocol(*rule.Protocols)
+		singleRule["destination_port"] = strings.ToLower(rule.DestinationPortRange)
+		singleRule["destination_ip"] = strings.ToLower(rule.DestinationIP)
+		if rule.DestinationVM != nil {
+			singleRule["destination_vm_id"] = getVmIdFromVmVappLocalId(vapp, rule.DestinationVM.VAppScopedVMID)
+			singleRule["destination_vm_nic_id"] = rule.DestinationVM.VMNicID
+			singleRule["destination_vm_ip_type"] = rule.DestinationVM.IPType
+		}
+		singleRule["source_port"] = strings.ToLower(rule.SourcePortRange)
+		singleRule["source_ip"] = strings.ToLower(rule.SourceIP)
+		if rule.SourceVM != nil {
+			singleRule["source_vm_id"] = getVmIdFromVmVappLocalId(vapp, rule.SourceVM.VAppScopedVMID)
+			singleRule["source_vm_nic_id"] = rule.SourceVM.VMNicID
+			singleRule["source_vm_ip_type"] = rule.SourceVM.IPType
+		}
+		singleRule["enable_logging"] = rule.EnableLogging
+		rules = append(rules, singleRule)
+	}
+	_ = d.Set("rule", rules)
+	_ = d.Set("default_action", vappNetwork.Configuration.Features.FirewallService.DefaultAction)
+	_ = d.Set("log_default_action", vappNetwork.Configuration.Features.FirewallService.LogDefaultAction)
+
+	return nil
+}
+
+func getVmIdFromVmVappLocalId(vapp *govcd.VApp, vmVappLocalId string) string {
```

Comment:
What is the purpose of this function? The naming doesn't help (what's a `vmVapp`? and a `LocalId` ?)

Reply:
This is some different ID which is required for mapping. It's called `VmVappLocalId`. API things, not my namings.

Reply:
But I can't ask user that strange ID to be provided by user, so I work with two ID's provided by user and needed for API.

Reply:
Would be great to have a function comment explaining what is this `VAppScopedLocalID` and where does it come from.